### ISSUE TITLE
Low pseudo-alignment % for CITE-seq data

### ISSUE THREAD

**Issue by tilschaef** [pachterlab/kb_python#149]
**Describe the issue**
Dear all,

I have used the kb kite workflow to quantify ADTs from CITE-seq data. It appears that there is a very low percentage of reads pseudo-aligned across multiple samples. In the specific case of this sample, only 15.3% of the initial 16651436 reads were pseudo-aligned. The picture is similar for the remaining samples.  However, when analyzing the same data with CITE-seq-Count, around 95% of the reads were aligned with the same barcode/UMI layout. 

I am a bit puzzled as to where this substantial difference may stem from. I have tried different k-mer lengths for building the index and/or included multi mappers which did not seem to increase the pseudo-alignment rate. 

I would appreciate any tips or feedback to point me in the right direction.

Cheers!

**What is the exact command that was run?**
```
whitelist="/scratch/etanis/refs/barcode_384_1_column.tab"
kb ref fb.tsv --workflow kite --verbose -i fb.idx -g fb_t2g.txt -f1 fb_cdna.fa >kb_ref.log 2>&1
kb count -i fb.idx -g fb_t2g.txt -w ${whitelist} --workflow kite \
         -x 0,8,16:0,0,8:1,0,0 --overwrite --verbose -t 40 \
          -o kb_out_INV_ADT1636 INV_ADT1636_R1.fastq.gz INV_ADT1636_R2.fastq.gz >kb_count.log 2>&1
```

**Command output (with `--verbose` flag)**
```
2021-11-05 13:56:02,373]   DEBUG [main] Printing verbose output
[2021-11-05 13:56:02,373]   DEBUG [main] kallisto binary located at /scratch/tschaefers/envs/kb/bin/kallisto
[2021-11-05 13:56:02,373]   DEBUG [main] bustools binary located at /scratch/tschaefers/envs/kb/bin/bustools
[2021-11-05 13:56:02,373]   DEBUG [main] Creating `kb_out_INV_ADT1636/tmp` directory
[2021-11-05 13:56:02,373]   DEBUG [main] Namespace(list=False, command='count', tmp=None, keep_tmp=False, verbose=True, kallisto='kallisto', bustools='bustools', i='../ref/fb.idx', g='../ref/fb_t2g.txt', x='0,8,16:0,0,8:1,0,0', o='kb_out_INV_ADT1636', w='/scratch/etanis/refs/barcode_384_1_column.tab', t=40, m='4G', workflow='kite', mm=False, tcc=False, filter=None, c1=None, c2=None, overwrite=True, dry_run=False, loom=False, h5ad=False, cellranger=False, report=False, no_inspect=False, no_validate=False, fastqs=['INV_ADT1636_R1.fastq.gz', 'INV_ADT1636_R2.fastq.gz'])
[2021-11-05 13:56:02,387]    INFO [count] Using index ../ref/fb.idx to generate BUS file to kb_out_INV_ADT1636 from
[2021-11-05 13:56:02,387]    INFO [count]         INV_ADT1636_R1.fastq.gz
[2021-11-05 13:56:02,387]    INFO [count]         INV_ADT1636_R2.fastq.gz
[2021-11-05 13:56:02,387]   DEBUG [count] kallisto bus -i ../ref/fb.idx -o kb_out_INV_ADT1636 -x 0,8,16:0,0,8:1,0,0 -t 40 INV_ADT1636_R1.fastq.gz INV_ADT1636_R2.fastq.gz
[2021-11-05 13:56:26,090]   DEBUG [count]
[2021-11-05 13:56:26,090]   DEBUG [count] [index] k-mer length: 11
[2021-11-05 13:56:26,090]   DEBUG [count] [index] number of targets: 2,788
[2021-11-05 13:56:26,090]   DEBUG [count] [index] number of k-mers: 2,788
[2021-11-05 13:56:26,090]   DEBUG [count] [index] number of equivalence classes: 2,788
[2021-11-05 13:56:26,090]   DEBUG [count] [quant] will process sample 1: INV_ADT1636_R1.fastq.gz
[2021-11-05 13:56:26,090]   DEBUG [count] INV_ADT1636_R2.fastq.gz
[2021-11-05 13:56:26,090]   DEBUG [count] [quant] finding pseudoalignments for the reads ... done
[2021-11-05 13:56:26,090]   DEBUG [count] [quant] processed 16,651,436 reads, 2,548,241 reads pseudoaligned
[2021-11-05 13:56:26,090]   DEBUG [count]
[2021-11-05 13:56:26,241]   DEBUG [count] kb_out_INV_ADT1636/output.bus passed validation
[2021-11-05 13:56:26,241]    INFO [count] Sorting BUS file kb_out_INV_ADT1636/output.bus to kb_out_INV_ADT1636/tmp/output.s.bus
[2021-11-05 13:56:26,242]   DEBUG [count] bustools sort -o kb_out_INV_ADT1636/tmp/output.s.bus -T kb_out_INV_ADT1636/tmp -t 40 -m 4G kb_out_INV_ADT1636/output.bus
[2021-11-05 13:56:28,375]   DEBUG [count] Read in 2548241 BUS records
[2021-11-05 13:56:28,400]   DEBUG [count] kb_out_INV_ADT1636/tmp/output.s.bus passed validation
[2021-11-05 13:56:28,400]    INFO [count] Inspecting BUS file kb_out_INV_ADT1636/tmp/output.s.bus
[2021-11-05 13:56:28,400]   DEBUG [count] bustools inspect -o kb_out_INV_ADT1636/inspect.json -w /scratch/etanis/refs/barcode_384_1_column.tab -e kb_out_INV_ADT1636/matrix.ec kb_out_INV_ADT1636/tmp/output.s.bus
[2021-11-05 13:56:28,455]    INFO [count] Correcting BUS records in kb_out_INV_ADT1636/tmp/output.s.bus to kb_out_INV_ADT1636/tmp/output.s.c.bus with whitelist /scratch/etanis/refs/barcode_384_1_column.tab
[2021-11-05 13:56:28,456]   DEBUG [count] bustools correct -o kb_out_INV_ADT1636/tmp/output.s.c.bus -w /scratch/etanis/refs/barcode_384_1_column.tab kb_out_INV_ADT1636/tmp/output.s.bus
[2021-11-05 13:56:28,533]   DEBUG [count] Found 384 barcodes in the whitelist
[2021-11-05 13:56:28,533]   DEBUG [count] Processed 1004102 BUS records
[2021-11-05 13:56:28,533]   DEBUG [count] In whitelist = 912393
[2021-11-05 13:56:28,533]   DEBUG [count] Corrected    = 46055
[2021-11-05 13:56:28,533]   DEBUG [count] Uncorrected  = 45654
[2021-11-05 13:56:28,555]   DEBUG [count] kb_out_INV_ADT1636/tmp/output.s.c.bus passed validation
[2021-11-05 13:56:28,555]    INFO [count] Sorting BUS file kb_out_INV_ADT1636/tmp/output.s.c.bus to kb_out_INV_ADT1636/output.unfiltered.bus
[2021-11-05 13:56:28,555]   DEBUG [count] bustools sort -o kb_out_INV_ADT1636/output.unfiltered.bus -T kb_out_INV_ADT1636/tmp -t 40 -m 4G kb_out_INV_ADT1636/tmp/output.s.c.bus
[2021-11-05 13:56:30,156]   DEBUG [count] Read in 958448 BUS records
[2021-11-05 13:56:30,177]   DEBUG [count] kb_out_INV_ADT1636/output.unfiltered.bus passed validation
[2021-11-05 13:56:30,177]    INFO [count] Generating count matrix kb_out_INV_ADT1636/counts_unfiltered/cells_x_features from BUS file kb_out_INV_ADT1636/output.unfiltered.bus
[2021-11-05 13:56:30,177]   DEBUG [count] bustools count -o kb_out_INV_ADT1636/counts_unfiltered/cells_x_features -g ../ref/fb_t2g.txt -e kb_out_INV_ADT1636/matrix.ec -t kb_out_INV_ADT1636/transcripts.txt --genecounts kb_out_INV_ADT1636/output.unfiltered.bus
[2021-11-05 13:56:30,277]   DEBUG [count] kb_out_INV_ADT1636/counts_unfiltered/cells_x_features.mtx passed validation
[2021-11-05 13:56:30,278]   DEBUG [main] Removing `kb_out_INV_ADT1636/tmp` directory

```

**Comment by sbooeshaghi** [comment]
Could you share the list of ADTs you are aligning against? My guess is that there is a long subsequence that is shared by two or  more of these tags.

**Comment by tilschaef** [comment]
Thank you for you quick response, I have attached a file with the ADTs to this reply.
[fb.txt](https://github.com/pachterlab/kb_python/files/7495933/fb.txt)
.

**Comment by tilschaef** [comment]
@sbooeshaghi , the problem seemed to be related to the constant base that was added to each ADT sequence before building the index. Initially, I added a constant base at the beginning of each sequence since all tag sequences are of length 10. However, appending it to the end instead yielded pseudo_alignment rates of ~90% and comparable results with CITESeq-count (after some filtering).  

In the case of this analysis, where the ADT is at the beginning of each read, it appears to matter on which site of the ADT the constant base is added. Lastly, adding a G after each ADT instead of the other bases yields the lowest alignment rates of ~75% which is something that I did not expect to make such a difference.
